bfdd/bfd.c

```diff
@@ -317,8 +328,9 @@ void ptm_bfd_sess_up(struct bfd_session *bfd)
 	}
 }
 
-void ptm_bfd_sess_dn(struct bfd_session *bfd, uint8_t diag)
+void ptm_bfd_sess_dn(struct bfd_session *bfd, uint8_t diag, uint8_t peer_state)
```

Comment:
Please, don't change this function signature. Don't you already have that new information in `struct bfd_session`?

Copied from: https://github.com/FRRouting/frr/pull/5002#discussion_r325662819

Reply:
Please have a look at line no: 887, neighbor state is being passed here, which is not available in struct bfd_session
To handle this issue, neighbor state is required ..

Explaining the scenario where this code change is required:
1. BFD is un-configured from BGP at remote end.
      - Neighbour BFD sends ADMIN_DOWN state, but BFD on local side will send DOWN to BGP, resulting in BGP 
         session DOWN. 
     -  Removing BFD session administratively shouldn't bring DOWN BGP session at local or remote.

2. BFD is un-configured from BGP or shutdown locally.
    -  BFD will send state DOWN to BGP resulting in BGP  session DOWN. (This is akin to saying do not use BFD 
        for BGP)
    -  Removing BFD session administratively shouldn't bring DOWN BGP session at local or remote.

Reply:
Nice catch, you are correct about the problem.

However I'd suggest to fix this differently: instead of changing old functions parameters, lets create a new `administrative down` function to handle this situation properly. Today we can only see this problem, but tomorrow we might find more exceptions for this case (I'd go further and create a administrative session down counter).

You don't need to follow the old naming patterns, those were used in the original Cumulus PTM daemon. I'd pick something like `bs_admin_down_handler` (bs -> bfd session) or `bfd_sess_admin_down`.

Reply:
Sure we can create a new function as you suggested "bs_admin_down_handler", but still we will have to change prototype of "control_notify". Let me know if it sounds good, i can work on modifying code accordingly.

Reply:
@rzalamena : Would appreciate your comments here